### PR TITLE
phylactery: init at 0.1.1

### PR DESCRIPTION
**Pull request by McSinyx**
###### Description of changes

Phylactery is a simple web app rendering comics from CBZ archives.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/phylactery/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{ lib, stdenv, fetchFromSourcehut, go }:
+
+stdenv.mkDerivation rec {
```
Comment (ckiee) [inline pkgs/servers/phylactery/default.nix:3]: Use `buildGoModule`, see [the docs](https://nixos.org/manual/nixpkgs/unstable/#sec-language-go).

File: nixos/modules/services/web-apps/phylactery.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+with lib;
+let cfg = config.services.phylactery;
+in {
+  options.services.phylactery = {
+    enable = mkEnableOption "Whether to enable Phylactery server.";
+
+    host = mkOption {
+      type = types.str;
+      default = "";
```
Comment (ckiee) [inline nixos/modules/services/web-apps/phylactery.nix:11]: Either remove the default or make the type `types.nullOr types.str` and make the default null.

File: nixos/modules/services/web-apps/phylactery.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        PHYLACTERY_ADDRESS = "${cfg.host}:${toString cfg.port}";
+        PHYLACTERY_LIBRARY = cfg.library;
+      };
+
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ConditionPathExists = cfg.library;
+        DynamicUser = true;
+        ExecStart = "${pkgs.phylactery}/bin/phylactery";
```
Comment (ckiee) [inline nixos/modules/services/web-apps/phylactery.nix:38]: Adding a `package` option with this as the default would be nice.

File: nixos/modules/services/web-apps/phylactery.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    systemd.services.phylactery = {
+      environment = {
+        PHYLACTERY_ADDRESS = "${cfg.host}:${toString cfg.port}";
+        PHYLACTERY_LIBRARY = cfg.library;
+      };
+
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ConditionPathExists = cfg.library;
```
Comment (ckiee) [inline nixos/modules/services/web-apps/phylactery.nix:36]: What does it do if it isn't around?

File: nixos/tests/phylactery.nix [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
```
Comment (ckiee) [inline nixos/tests/phylactery.nix:1]: Separate this into its own commit. `nixos/tests: add phlactery`

**Review by ckiee — CHANGES_REQUESTED** [review]
Looking pretty nice.

File: nixos/modules/services/web-apps/phylactery.nix (hunk 2 above)
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix:11]: The implementation allows an empty string for host, e.g. :8080 is considered a valid address.  If localhost is better, feel free to tell me to change.

File: nixos/modules/services/web-apps/phylactery.nix (hunk 4 above)
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix:36]: Everything would respond as 404 so this check is nice to have IMHO.

File: nixos/modules/services/web-apps/phylactery.nix (hunk 3 above)
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix:38]: Could you please elaborate on the reason?

File: nixos/tests/phylactery.nix (hunk 5 above)
Comment (McSinyx) [reply nixos/tests/phylactery.nix:1]: Done!

File: nixos/modules/services/web-apps/phylactery.nix (hunk 2 above)
Comment (ckiee) [reply nixos/modules/services/web-apps/phylactery.nix:11]: `localhost` would be clearer.
```suggestion
      type = types.str;
      default = "localhost";
```

File: nixos/modules/services/web-apps/phylactery.nix (hunk 3 above)
Comment (ckiee) [reply nixos/modules/services/web-apps/phylactery.nix:38]: Well, if I wanted to override `pkgs.phylactery` I'd have to use a nixpkgs overlay, and that can effect other unrelated usages of `phylactery`. If instead I could set `config.services.phylactery.package` that would make it contained to `nixos/phylactery`.

Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix]: Thanks, done!

File: nixos/modules/services/web-apps/phylactery.nix (hunk 2 above)
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix:11]: Applied!

File: pkgs/servers/phylactery/default.nix (hunk 1 above)
Comment (McSinyx) [reply pkgs/servers/phylactery/default.nix]: That works, thanks!

**Review by ckiee — CHANGES_REQUESTED** [review]

File: pkgs/servers/web-apps/phylactery/default.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  goPackagePath = "git.sr.ht/~cnx/phylactery";
+
+  src = fetchFromSourcehut {
+    owner = "~cnx";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-HcpdPQ1WOMFNdfsZb//GvUVBCbmS3jARbcXjchlv2oE=";
+  };
+
+  goDeps = null;
```
[the remark this answers is not in the document]
Comment (McSinyx) [reply pkgs/servers/web-apps/phylactery/default.nix:16]: Thanks, didn't know that, probably not worth flexing though :smile:

File: nixos/modules/services/web-apps/phylactery.nix [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      description = "Listen host for Phylactery.";
+    };
+
+    port = mkOption {
+      type = types.port;
+      description = "Listen port for Phylactery.";
+    };
+
+    library = mkOption {
+      type = types.str;
```
[the remark this answers is not in the document]
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix]: Thanks, applied.

File: nixos/modules/services/web-apps/phylactery.nix [hunk 8]
```diff
@@ -0,0 +1,51 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let cfg = config.services.phylactery;
+in {
+  options.services.phylactery = {
+    enable = mkEnableOption "Whether to enable Phylactery server.";
```
[the remark this answers is not in the document]
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix]: Thanks, done!

File: nixos/modules/services/web-apps/phylactery.nix [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      defaultText = literalExpression "pkgs.phylactery";
+      description = "The Phylactery package to use.";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.phylactery = {
+      environment = {
+        PHYLACTERY_ADDRESS = "${cfg.host}:${toString cfg.port}";
+        PHYLACTERY_LIBRARY = cfg.library;
```
[the remark this answers is not in the document]
Comment (McSinyx) [reply nixos/modules/services/web-apps/phylactery.nix]: Explicit is better than implicit I guess :shrug:

File: nixos/tests/web-apps/phylactery.nix [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+      enable = true;
+      port = 8080;
+      library = "/tmp";
+    };
+  };
+
+  testScript = ''
+    start_all()
+    machine.wait_for_unit("phylactery")
+    machine.wait_for_open_port(8080)
```
[the remark this answers is not in the document]
Comment (McSinyx) [reply nixos/tests/web-apps/phylactery.nix:15]: I added `/` fetching, and HTTP is beyond TCP I suppose, but not much meaningful stuff can really be done here.  I'll add tests to upstream in later releases.

File: pkgs/servers/web-apps/phylactery/default.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (ckiee) [reply pkgs/servers/web-apps/phylactery/default.nix]: I'm incredibly literal and impersonal in code review usually, wasn't flexing.

**Review by ckiee — APPROVED** [review]

File: nixos/modules/services/web-apps/phylactery.nix [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.phylactery;
+      defaultText = literalExpression "pkgs.phylactery";
+      description = "The Phylactery package to use";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
```
Comment (ckiee) [inline nixos/modules/services/web-apps/phylactery.nix:33]: Only noticed in my final pass
```suggestion
  config = mkIf cfg.enable {
```

File: pkgs/servers/web-apps/phylactery/default.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (McSinyx) [reply pkgs/servers/web-apps/phylactery/default.nix]: No I was referring to flexing *the lack of dependencies* :grin:
